src/dead_letters.rs

```diff
@@ -0,0 +1,255 @@
+use crate::transforms::Transformer;
+use async_trait::async_trait;
+use chrono::prelude::*;
+use core::fmt::Debug;
+use log::{error, info, warn};
+use parquet::errors::ParquetError;
+use rdkafka::message::BorrowedMessage;
+use serde::{Deserialize, Serialize};
+use serde_json::Value;
+use std::collections::HashMap;
+
+use crate::{deltalake_ext::*, transforms::TransformError};
+
+/// Struct that represents a dead letter record.
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct DeadLetter {
+    /// Base64 encoded bytes captured when message deserialization from Kafka fails.
+    pub base64_bytes: Option<String>,
+    /// JSON string captured when either transform or parquet write fails.
+    pub json_string: Option<String>,
+    /// Error string that correlates with the failure.
+    /// The error type will vary depending on the context.
+    pub error: Option<String>,
+    /// Timestamp microseconds representing when the dead letter was created.
+    /// Microseconds are used for compatability with the deltalake `timestamp` type.
+    pub timestamp: i64,
+}
+
+impl DeadLetter {
+    /// Creates a dead letter from bytes that failed deserialization.
+    /// `json_string` will always be `None`.
+    pub fn from_failed_deserialization(bytes: &[u8], err: serde_json::Error) -> Self {
+        let timestamp = Utc::now();
+        Self {
+            base64_bytes: Some(base64::encode(bytes)),
+            json_string: None,
+            error: Some(err.to_string()),
+            timestamp: timestamp.timestamp_nanos() / 1000,
+        }
+    }
+
+    /// Creates a dead letter from a failed transform.
+    /// `base64_bytes` will always be `None`.
+    pub fn from_failed_transform(value: &Value, err: TransformError) -> Self {
+        let timestamp = Utc::now();
+        match serde_json::to_string(value) {
```

Comment:
if you call `(v: Value).to_string()` instead you won't have to bother with pattern matching and `unreachable!()` calls

Reply:
nice catch - fixed